packages/messaging-amqp/src/AmqpModule.ts

```diff
@@ -0,0 +1,415 @@
+/*
+ * © Copyright 2022 HP Development Company, L.P.
+ * SPDX-License-Identifier: MIT
+ */
+
+import {
+	App,
+	executeInterceptorsStack,
+	getInterceptorsHandlers,
+	InterceptorBag,
+	mapParallel,
+	mapSeries,
+	Module,
+	nextTick
+} from '@davinci/core';
+import { ClassReflection, ClassType, DecoratorId, MethodReflection, PartialDeep } from '@davinci/reflector';
+import { Level, Logger, pino } from 'pino';
+import createDeepmerge from '@fastify/deepmerge';
+import amqplib, { Channel } from 'amqplib';
+import amqpConnectionManager, { AmqpConnectionManager, ChannelWrapper } from 'amqp-connection-manager';
+import stringify from 'fast-json-stable-stringify';
+import { EventEmitter } from 'events';
+import { AmqpSubscriptionSettings, ParameterConfiguration, SubscribeDecoratorMetadata, Subscription } from './types';
+
+const deepmerge = createDeepmerge({ all: true });
+
+export interface AmqpModuleOptions {
+	connection: string | amqplib.Options.Connect;
+	connectionTimeout?: number;
+	subscriptions?: Array<AmqpSubscriptionSettings>;
+	defaultSubscriptionSettings?: PartialDeep<AmqpSubscriptionSettings>;
+	gracefulShutdownStrategy?: 'none' | 'processInFlight' | 'nackInFlight' | null;
+	logger?: {
+		name?: string;
+		level?: Level | 'silent';
+	};
+}
+
+export class AmqpModule extends Module {
+	app: App;
+	private logger: Logger;
+	private options: AmqpModuleOptions;
+	private connection: AmqpConnectionManager;
+	private subscriptions: Array<Subscription>;
+	private subscriptionHash: Record<string, Subscription> = {};
+	private channelsHash: Record<string, ChannelWrapper> = {};
+	private bus: EventEmitter;
+	private totalInFlight: number;
+
+	constructor(options: AmqpModuleOptions) {
+		super();
+		this.options = deepmerge({ connectionTimeout: 5000, logger: { name: 'AmqpModule', level: 'info' } }, options);
+		this.bus = new EventEmitter();
+		this.logger = pino({ name: this.options.logger?.name });
+		this.logger.level = this.options.logger?.level;
+		this.totalInFlight = 0;
+	}
+
+	getModuleId() {
+		return ['messaging', 'amqp'];
+	}
+
+	onRegister(app: App) {
+		this.app = app;
+		const connectionOptions =
+			typeof this.options.connection === 'object' ? this.options.connection : { url: this.options.connection };
+		this.connection = amqpConnectionManager.connect(connectionOptions);
```

Comment:
is there any way to specify heartbeatIntervalInSeconds / reconnectTimeInSeconds ? I think we are using those ones in messaging lib.